controller/controller.go

```diff
@@ -0,0 +1,520 @@
+	recorder.StartRecordingToSink(&corev1.EventSinkImpl{Interface: c.kubeClient.CoreV1().Events(ns)})
+	eRecorder := recorder.NewRecorder(scheme.Scheme, v1.EventSource{Component: leader})
+
+	rlConfig := resourcelock.ResourceLockConfig{
+		Identity:      sanitize(id),
+		EventRecorder: eRecorder,
+	}
+
+	l, err := resourcelock.New(resourcelock.LeasesResourceLock, ns, leader, c.kubeClient.CoreV1(), c.kubeClient.CoordinationV1(), rlConfig)
+	if err != nil {
+		return err
+	}
+
+	leaderConfig := leaderelection.LeaderElectionConfig{
+		Lock:          l,
+		LeaseDuration: c.LeaseDuration,
+		RenewDeadline: c.RenewDeadline,
+		RetryPeriod:   c.RetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				glog.V(2).Info("became leader, starting")
+				c.runController(ctx)
+			},
+			OnStoppedLeading: func() {
+				glog.Fatal("stopped leading")
+			},
+			OnNewLeader: func(identity string) {
+				glog.V(3).Infof("new leader detected, current leader: %s", identity)
+			},
+		},
+	}
+
+	leaderelection.RunOrDie(ctx, leaderConfig)
+	return nil // should never reach here
+}
+
+func (c *ObjectStorageController) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *ObjectStorageController) processNextItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	op, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	// With the lock below in place, we can safely tell the queue that we are done
+	// processing this item. The lock will ensure that multiple items of the same
+	// name and kind do not get processed simultaneously
+	defer c.queue.Done(op)
+
+	// Ensure that multiple operations on different versions of the same object
+	// do not happen in parallel
+	c.OpLock(op)
+	defer c.OpUnlock(op)
+
+	var err error
+	switch o := op.(type) {
+	case addOp:
+		add := *o.AddFunc
+		err = add(ctx, o.Object)
+	case updateOp:
+		update := *o.UpdateFunc
+		err = update(ctx, o.OldObject, o.NewObject)
+	case deleteOp:
+		delete := *o.DeleteFunc
+		err = delete(ctx, o.Object)
+	default:
+		panic("unknown item in queue")
+	}
+
+	// Handle the error if something went wrong
+	c.handleErr(err, op)
+	return true
+}
+
+func (c *ObjectStorageController) OpLock(op interface{}) {
+	c.GetOpLock(op).Lock()
+}
+
+func (c *ObjectStorageController) OpUnlock(op interface{}) {
+	c.GetOpLock(op).Unlock()
+}
+
+func (c *ObjectStorageController) GetOpLock(op interface{}) *sync.Mutex {
+	var key string
+	var ext string
+
+	switch o := op.(type) {
+	case addOp:
+		key = o.Key
+		ext = fmt.Sprintf("%v", o.AddFunc)
+	case updateOp:
+		key = o.Key
+		ext = fmt.Sprintf("%v", o.UpdateFunc)
+	case deleteOp:
+		key = o.Key
+		ext = fmt.Sprintf("%v", o.DeleteFunc)
+	default:
+		panic("unknown item in queue")
+	}
+
+	lockKey := fmt.Sprintf("%s/%s", key, ext)
+	if c.locker == nil {
+		c.locker = map[string]*sync.Mutex{}
+	}
+	if _, ok := c.locker[lockKey]; !ok {
+		c.lockerLock.Lock()
```

Comment:
We need to acquire this lock outside the if-condition. When there are multiple requests for the same key (that does not exist yet), they could race and lead to concurrent map access panic. 


This should be outside the if-block
```
c.lockerLock.Lock()
defer c.lockerLock.Unlock()
```

Reply:
Fixed